File: caregiver_dashboard/dashboard_backend/routes/notifications.js
const express = require('express');
const Notification = require('../models/Notification');
const { protect } = require('../middleware/auth');

const router = express.Router();
router.use(protect);

// GET /api/notifications  — get all notifications for current user
router.get('/', async (req, res) => {
  try {
    const { unread_only, limit = 30 } = req.query;
    const query = { recipient_id: req.user._id, is_dismissed: { $ne: true } };
    if (unread_only === 'true') query.is_read = false;

    const notifications = await Notification.find(query)
      .sort({ createdAt: -1 })
      .limit(parseInt(limit));

    const unread_count = await Notification.countDocuments({ recipient_id: req.user._id, is_read: false, is_dismissed: { $ne: true } });
    res.json({ notifications, unread_count });
  } catch (err) { res.status(500).json({ error: err.message }); }
});

// PATCH /api/notifications/:id/read  — mark as read
router.patch('/:id/read', async (req, res) => {
  try {
    const notification = await Notification.findOneAndUpdate(
      { _id: req.params.id, recipient_id: req.user._id },
      { is_read: true, read_at: new Date() },
      { new: true }
    );
    if (!notification) return res.status(404).json({ error: 'Notification not found' });
    res.json(notification);
  } catch (err) { res.status(500).json({ error: err.message }); }
});

// PATCH /api/notifications/read-all  — mark all as read
router.patch('/read-all', async (req, res) => {
  try {
    await Notification.updateMany(
      { recipient_id: req.user._id, is_read: false },
      { is_read: true, read_at: new Date() }
    );
    res.json({ message: 'All notifications marked as read' });
  } catch (err) { res.status(500).json({ error: err.message }); }
});

// PATCH /api/notifications/:id/acknowledge  — acknowledge an alert
router.patch('/:id/acknowledge', async (req, res) => {
  try {
    const notification = await Notification.findOneAndUpdate(
      { _id: req.params.id, recipient_id: req.user._id },
      { acknowledged_at: new Date(), is_read: true, read_at: new Date() },
      { new: true }
    );
    if (!notification) return res.status(404).json({ error: 'Notification not found' });
    res.json(notification);
  } catch (err) { res.status(500).json({ error: err.message }); }
});

// DELETE /api/notifications/:id  — dismiss
router.delete('/:id', async (req, res) => {
  try {
    const notification = await Notification.findOneAndUpdate(
      { _id: req.params.id, recipient_id: req.user._id },
      { is_dismissed: true },
      { new: true }
    );
    if (!notification) return res.status(404).json({ error: 'Notification not found' });
    res.json({ message: 'Notification dismissed' });
  } catch (err) { res.status(500).json({ error: err.message }); }
});

module.exports = router;

File: medication_backend/routes/notifications.py
from fastapi import APIRouter, Depends, Query, HTTPException
from typing import List, Optional
from bson import ObjectId
from database import get_db
from routes.auth import get_current_user

router = APIRouter(prefix="/api/notifications", tags=["notifications"])

def serialize_notif(notif: dict) -> dict:
    if "_id" in notif:
        notif["id"] = str(notif.pop("_id"))
        notif["_id"] = notif["id"]
    for key in ["recipient_id", "subject_user_id"]:
        if key in notif and isinstance(notif[key], ObjectId):
            notif[key] = str(notif[key])
    for key in ["createdAt", "updatedAt", "read_at", "acknowledged_at", "created_at", "updated_at"]:
        if key in notif and notif[key]:
            notif[key] = notif[key].isoformat() + "Z"
    return notif

@router.get("/")
async def get_notifications(
    limit: int = Query(30),
    unread_only: bool = Query(False),
    db=Depends(get_db),
    current_user=Depends(get_current_user)
):
    query = {
        "recipient_id": ObjectId(current_user["id"]),
        "is_dismissed": {"$ne": True}
    }
    if unread_only:
        query["is_read"] = False

    cursor = db.notifications.find(query).sort("createdAt", -1).limit(limit)
    notifications = await cursor.to_list(length=limit)

    unread_count = await db.notifications.count_documents({
        "recipient_id": ObjectId(current_user["id"]),
        "is_read": False,
        "is_dismissed": {"$ne": True}
    })

    serialized = [serialize_notif(n) for n in notifications]
    return {"notifications": serialized, "unread_count": unread_count}

@router.patch("/{notif_id}/read")
async def mark_read(notif_id: str, db=Depends(get_db), current_user=Depends(get_current_user)):
    from datetime import datetime
    result = await db.notifications.update_one(
        {"_id": ObjectId(notif_id), "recipient_id": ObjectId(current_user["id"])},
        {"$set": {"is_read": True, "read_at": datetime.utcnow()}}
    )
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"message": "Marked as read"}

@router.patch("/read-all")
async def mark_all_read(db=Depends(get_db), current_user=Depends(get_current_user)):
    from datetime import datetime
    await db.notifications.update_many(
        {"recipient_id": ObjectId(current_user["id"]), "is_read": False},
        {"$set": {"is_read": True, "read_at": datetime.utcnow()}}
    )
    return {"message": "All notifications marked as read"}

@router.patch("/{notif_id}/acknowledge")
async def acknowledge_notification(notif_id: str, db=Depends(get_db), current_user=Depends(get_current_user)):
    from datetime import datetime
    result = await db.notifications.update_one(
        {"_id": ObjectId(notif_id), "recipient_id": ObjectId(current_user["id"])},
        {"$set": {"acknowledged_at": datetime.utcnow(), "is_read": True, "read_at": datetime.utcnow()}}
    )
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"message": "Notification acknowledged"}

@router.delete("/{notif_id}")
async def dismiss_notification(notif_id: str, db=Depends(get_db), current_user=Depends(get_current_user)):
    result = await db.notifications.update_one(
        {"_id": ObjectId(notif_id), "recipient_id": ObjectId(current_user["id"])},
        {"$set": {"is_dismissed": True}}
    )
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"message": "Notification dismissed"}
